internal/service/appstream/stack_user_association.go

```diff
@@ -0,0 +1,177 @@
+package appstream
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appstream"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+)
+
+func ResourceStackUserAssociation() *schema.Resource {
+	return &schema.Resource{
+		CreateWithoutTimeout: resourceStackUserAssociationCreate,
+		ReadWithoutTimeout:   resourceStackUserAssociationRead,
+		DeleteWithoutTimeout: resourceStackUserAssociationDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"authentication_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(appstream.AuthenticationType_Values(), false),
+			},
+			"send_email_notification": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+			"stack_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 128),
+			},
+			"user_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceStackUserAssociationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).AppStreamConn
+
+	input := &appstream.UserStackAssociation{
+		AuthenticationType: aws.String(d.Get("authentication_type").(string)),
+		StackName:          aws.String(d.Get("stack_name").(string)),
+		UserName:           aws.String(d.Get("user_name").(string)),
+	}
+
+	if v, ok := d.GetOk("send_email_notification"); ok {
+		input.SendEmailNotification = aws.Bool(v.(bool))
+	}
+
+	var err error
+	var output *appstream.BatchAssociateUserStackOutput
+	err = resource.RetryContext(ctx, fleetOperationTimeout, func() *resource.RetryError {
+		output, err = conn.BatchAssociateUserStackWithContext(ctx, &appstream.BatchAssociateUserStackInput{
+			UserStackAssociations: []*appstream.UserStackAssociation{input},
+		})
+		if err != nil {
+			if tfawserr.ErrCodeEquals(err, appstream.ErrCodeResourceNotFoundException) {
+				return resource.RetryableError(err)
+			}
+
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		output, err = conn.BatchAssociateUserStackWithContext(ctx, &appstream.BatchAssociateUserStackInput{
+			UserStackAssociations: []*appstream.UserStackAssociation{input},
+		})
+	}
```

Comment:
Does this need a `Retry` loop? The documentation for [`BatchAssociateUserStack`](https://docs.aws.amazon.com/appstream2/latest/APIReference/API_BatchAssociateUserStack.html) seems to indicate that the function is synchronous.